tests/client/test_query.py

```diff
@@ -55,3 +56,65 @@ def test_get_container_sources():
+
+    # Set up a layered container image, a, which embeds an RPM and a Go executable.
+    #
+    # First, a parent image, p:
+    #   (p:Artifact) <-- (:Build) <-- (:SourceLocation)
+
+    # Then the container itself, a:
+    #   (a:Artifact) <-- (:Build) <-- (:SourceLocation)
+    #   (a:Artifact) <-[:EMBEDS]- (p:Artifact)
+    #
+    # And let's have the parent image embed an RPM, glibc:
+    #   (p:Artifact) <-[:EMBEDS]- (rpm:Artifact) <-- (:Build) <-- (internal:SourceLocation)
+    #                <-- (upstream:SourceLocation) <-- (:Component)
+
+    # Parent image, p
+    container_filename = ('docker-image-sha256:' + '0' * 64 + '.x86_64.tar.gz')
+    p = Artifact(
+        archive_id='742662', architecture='x86_64', filename=container_filename,
+        type_='container').save()
+    container_build_id = '1742662'
+    container_build = Build(id_=container_build_id, type_='container').save()
+    container_build.artifacts.connect(p)
+    container_internal_url = ('git://pks.domain.local/containers/p#...')
+    container_internal_source = SourceLocation(url=container_internal_url).save()
+    container_build.source_location.connect(container_internal_source)
+
+    # The container itself
+    container_filename = ('docker-image-sha256:' + '1' * 64 + '.x86_64.tar.gz')
+    a = Artifact(
+        archive_id='742663', architecture='x86_64', filename=container_filename,
+        type_='container').save()
+    container_build_id = '1742663'
+    container_build = Build(id_=container_build_id, type_='container').save()
+    container_build.artifacts.connect(a)
+    container_internal_url = ('git://pks.domain.local/containers/a#...')
+    container_internal_source = SourceLocation(url=container_internal_url).save()
+    container_build.source_location.connect(container_internal_source)
+    a.embedded_artifacts.connect(p)  # parent image
+
+    # glibc RPM for parent image
+    glibc_build = Build(id_='770188', type_='rpm').save()
+    glibc_rpm = Artifact(
+        archive_id='5818103', architecture='x86_64', filename='glibc-1.2.3-1.el7.x86_64.rpm',
+        type_='rpm').save()
+    glibc_build.artifacts.connect(glibc_rpm)
+    glibc_upstream_url = ('https://github.com/coreos/etcd/archive/.../glibc-1674e682.tar.gz')
+    glibc_upstream_source = SourceLocation(url=glibc_upstream_url).save()
+    glibc_internal_url = 'git://pks.domain.local/rpms/glibc#8fb38a89e1177b0303c675d206f90f6a83e2'
+    glibc_internal_source = SourceLocation(url=glibc_internal_url).save()
+    glibc_build.source_location.connect(glibc_internal_source)
+    glibc_internal_source.upstream.connect(glibc_upstream_source)
+    glibc_component = Component(canonical_type='generic', canonical_namespace='foo',
+                                canonical_name='bar')
+    glibc_upstream_source.component.connect(glibc_component)
+    p.embedded_artifacts.connect(glibc_rpm)
+
+    expected = []
+    assert query.get_source_components_in_archive('742663') == expected
```

Comment:
Could you add comment why do we expect empty list here?

Reply:
This was just work-in-progress data. Fixed now.